adapters/operaads/operaads.go

```diff
@@ -0,0 +1,232 @@
+	ErrEmptyEndpointId  = errors.New("empty endpoint id")
+	ErrEmptyPublisherId = errors.New("empty publisher id")
+)
+
+type OperaadsAdapter struct {
+	epTemplate *template.Template
+}
+
+// Builder builds a new instance of the operaads adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	epTemplate, err := template.New("endpoint").Parse(config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	bidder := &OperaadsAdapter{
+		epTemplate: epTemplate,
+	}
+	return bidder, nil
+}
+
+func (a *OperaadsAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	impCount := len(request.Imp)
+	requestData := make([]*adapters.RequestData, 0, impCount)
+	errs := []error{}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	err := checkRequest(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	for _, imp := range request.Imp {
+		requestCopy := *request
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		var operaadsExt openrtb_ext.ImpExtOperaads
+		if err := json.Unmarshal(bidderExt.Bidder, &operaadsExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if operaadsExt.PublisherID == "" {
+			errs = append(errs, ErrEmptyPublisherId)
+			continue
+		}
+		if operaadsExt.EndpointID == "" {
+			errs = append(errs, ErrEmptyEndpointId)
+			continue
+		}
+		if operaadsExt.PlacementID == "" {
+			errs = append(errs, ErrEmptyPlacementId)
+			continue
+		}
+
+		err := convertImpression(&imp)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		imp.TagID = operaadsExt.PlacementID
+
+		requestCopy.Imp = []openrtb2.Imp{imp}
+		reqJSON, err := json.Marshal(&requestCopy)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+
+		macro := macros.EndpointTemplateParams{PublisherID: operaadsExt.PublisherID, AccountID: operaadsExt.EndpointID}
+		endpoint, err := macros.ResolveMacros(*a.epTemplate, &macro)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		reqData := &adapters.RequestData{
+			Method:  http.MethodPost,
+			Uri:     endpoint,
+			Body:    reqJSON,
+			Headers: headers,
+		}
+		requestData = append(requestData, reqData)
+	}
+	return requestData, errs
+}
+
+func checkRequest(request *openrtb2.BidRequest) error {
+	if request.Device == nil || len(request.Device.OS) == 0 {
+		return &errortypes.BadInput{
+			Message: "Impression is missing device OS information",
+		}
+	}
+
+	return nil
+}
+
+func convertImpression(imp *openrtb2.Imp) error {
+	if imp.Banner == nil && imp.Video == nil && imp.Native == nil {
+		return &errortypes.BadInput{
+			Message: "Opera ads only supports banner, video or native ads.",
+		}
+	}
```

Comment:
This check is not necessary. It is performed by PBS-Core based on the formats listed in the bidder yaml file.

Reply:
Can I remove this code?

Reply:
Do you mean the bidder-info/operaads.yaml?

Reply:
Yes, you can remove it. PBS-Core verifies the request is of one of the formats specified in `bidder-info/operaads.yaml`.